core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java

```diff
@@ -140,56 +165,36 @@ private List<URIStatus> filterKeys(String bucketPrefix, List<URIStatus> children
     }
     //sort use uri path
     children.sort(Comparator.comparing(URIStatus::getPath));
-    return children.stream()
-        //marker filter
-        .filter((status) -> (status.getPath().compareTo(marker) > 0)
-            //prefix filter
-            && status.getPath().startsWith(bucketPrefix + mPrefix)
-            //startAfter filter for listObjectV2
-            && (!isVersion2() || status.getPath().compareTo(bucketPrefix + mStartAfter) > 0))
-        .limit(mMaxKeys)
-        .collect(Collectors.toList());
-  }
-
-  private void buildContentsAndCommonPrefix(String prefix, List<URIStatus> keys) {
-    final List<URIStatus> objectsList;
-    final List<URIStatus> prefixList;
-    if (mDelimiter == null) {
-      objectsList = keys;
-      prefixList = Collections.emptyList();
-    } else {
-      final Map<Boolean, List<URIStatus>> typeToStatus;
-      if (mDelimiter.equals(AlluxioURI.SEPARATOR)) {
-        typeToStatus = keys.stream()
-            .collect(Collectors.groupingBy(URIStatus::isFolder));
-      } else {
-        typeToStatus = keys.stream()
-            .collect(Collectors.groupingBy(
-                status -> status.getPath().substring(prefix.length()).contains(mDelimiter)
-            ));
-      }
-
-      objectsList = typeToStatus.getOrDefault(false, Collections.emptyList());
-      prefixList = typeToStatus.getOrDefault(true, Collections.emptyList());
-    }
-
-    mContents = new ArrayList<>();
-    for (URIStatus status : objectsList) {
-      mContents.add(new Content(
-          status.isFolder() ? status.getPath().substring(prefix.length())
-              + AlluxioURI.SEPARATOR : status.getPath().substring(prefix.length()),
-          S3RestUtils.toS3Date(status.getLastModificationTimeMs()),
-          status.isFolder() ? "0" : String.valueOf(status.getLength())
-      ));
-    }
-
-    mCommonPrefixes = new ArrayList<>();
-    for (URIStatus status : prefixList) {
-      final String path = status.getPath().substring(prefix.length());
-      if (mDelimiter.equals(AlluxioURI.SEPARATOR)) {
-        // "/" delimiter make sure prefix not repeat
-        mCommonPrefixes.add(new CommonPrefix(path + mDelimiter));
-      } else {
+    //group by common prefix if delimiter is provided
```

Comment:
Would it still be possible to do these things in a stream instead of a loop? Generally streams are cleaner.

Reply:
Yes but it involves multiple traversals in a row which I deemed to be not worth the cleanliness of stream processing.

I tried some combinations of filter/map/reduce but none of them were able to process the results in a single traversal. Also using a raw `for`-loop lets me `break` early whereas I couldn't find a way to pre-emptively exit a stream. If you know how please let me know though.

Reply:
Resolving as outdated